File: apps/api/app/schemas/builds.py
# apps/api/app/schemas/builds.py
from datetime import datetime
from typing import Any, Optional

from pydantic import AliasChoices, BaseModel, ConfigDict, Field


class VisionExtracted(BaseModel):
    """Extracted data from vision analysis."""
    make: Optional[str] = None
    model: Optional[str] = None
    year: Optional[str] = None
    confidence: Optional[int] = None
    part_name: Optional[str] = None
    specifications: Optional[dict[str, Any]] = None
    mods_detected: list[str] = []
    notes: Optional[str] = None


class VisionData(BaseModel):
    """Vision analysis result for an uploaded image."""
    image_type: str  # "car" | "rims" | "engine_bay" | "suspension" | "inspiration" | "part" | "unknown"
    summary: str
    extracted: VisionExtracted


class BuildCreate(BaseModel):
    """Shape of data the client sends when creating a build."""

    model_config = ConfigDict(populate_by_name=True)

    title: str
    car: Optional[str] = Field(
        default=None,
        validation_alias=AliasChoices("car", "donor_car"),
    )
    modification_goal: Optional[str] = None
    goals: list[str] = Field(default_factory=list)


class BuildUpdate(BaseModel):
    """Shape of data the client sends when partially updating a build."""

    model_config = ConfigDict(populate_by_name=True)

    title: Optional[str] = None
    car: Optional[str] = Field(
        default=None,
        validation_alias=AliasChoices("car", "donor_car"),
    )
    modification_goal: Optional[str] = None
    goals: Optional[list[str]] = None
    status: Optional[str] = None


class BuildResponse(BaseModel):
    """Shape of data returned to the client for list/create/update responses."""

    model_config = ConfigDict(populate_by_name=True)

    id: str
    user_id: str
    title: str
    car: Optional[str] = Field(
        default=None,
        validation_alias=AliasChoices("car", "donor_car"),
    )
    modification_goal: Optional[str] = None
    goals: list[str] = []
    image_url: Optional[str] = None
    status: str
    is_public: bool
    created_at: datetime
    updated_at: datetime
    parts_total: int = 0
    parts_sourced: int = 0
    vision_data: Optional[VisionData] = None


class PartVendorResponse(BaseModel):
    """Shape of a vendor option for a part."""

    model_config = ConfigDict(populate_by_name=True)

    id: str
    part_id: str
    vendor_name: str
    vendor_url: Optional[str] = None
    price: Optional[float] = None
    currency: str = "USD"
    ships_from: Optional[str] = None
    estimated_days_min: Optional[int] = None
    estimated_days_max: Optional[int] = None
    shipping_cost: Optional[float] = None
    is_primary: bool = False
    created_at: datetime


class PartVendorCreate(BaseModel):
    """Schema for creating a vendor option for a part."""

    model_config = ConfigDict(populate_by_name=True)

    vendor_name: str
    vendor_url: Optional[str] = None
    price: Optional[float] = None
    currency: str = "USD"
    ships_from: Optional[str] = None
    estimated_days_min: Optional[int] = None
    estimated_days_max: Optional[int] = None
    shipping_cost: Optional[float] = None
    is_primary: bool = False


class PartResponse(BaseModel):
    """Shape of a single part returned inside BuildDetailResponse."""

    model_config = ConfigDict(populate_by_name=True)

    id: str
    build_id: str
    name: str
    description: Optional[str] = None
    category: Optional[str] = None
    status: str
    price_estimate: Optional[float] = None
    vendor_url: Optional[str] = None
    image_url: Optional[str] = None
    is_safety_critical: bool = False
    notes: Optional[str] = None
    goal: Optional[str] = None
    vendors: list[PartVendorResponse] = []
    ordered_from_vendor_id: Optional[str] = None
    ordered_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class BuildDetailResponse(BaseModel):
    """Full build detail including all part data."""

    model_config = ConfigDict(populate_by_name=True)

    id: str
    user_id: str
    title: str
    car: Optional[str] = Field(
        default=None,
        validation_alias=AliasChoices("car", "donor_car"),
    )
    modification_goal: Optional[str] = None
    goals: list[str] = []
    image_url: Optional[str] = None
    status: str
    is_public: bool
    created_at: datetime
    updated_at: datetime
    parts: list[PartResponse] = []
    parts_total: int = 0
    parts_sourced: int = 0
    vision_data: Optional[VisionData] = None


class BuildImageResponse(BaseModel):
    """Shape of data returned to the client after uploading a build image."""
    image_url: str


class GenerateResponse(BaseModel):
    """Returned after a parts generation request."""
    parts_created: int
    build: BuildDetailResponse


class GeneratePartsRequest(BaseModel):
    """Request to generate parts for a build."""
    force_regenerate: bool = False


class GeneratePartsResponse(BaseModel):
    """Response after generating parts for a build."""
    build_id: str
    parts: list[PartResponse]
    total_parts: int
    estimated_total: float
    safety_critical_count: int
    message: str


class OrderPartRequest(BaseModel):
    """Request to order a part from a specific vendor."""
    vendor_id: str

File: apps/web/src/lib/api/builds.ts
// apps/web/src/lib/api/builds.ts
// Client-safe — no server-only imports. Server fetches live in builds.server.ts.
import type { Database } from "@/types/database"

type BuildRow = Database["public"]["Tables"]["builds"]["Row"]

export type Build = Omit<BuildRow, "donor_car" | "engine_swap"> & {
  car?: string | null
  donor_car?: string | null
  engine_swap?: string | null
  modification_goal?: string | null
  parts_total?: number
  parts_sourced?: number
}

export interface PartVendor {
  id: string
  part_id: string
  vendor_name: string
  vendor_url: string | null
  price: number | null
  currency: string
  ships_from: string | null
  estimated_days_min: number | null
  estimated_days_max: number | null
  shipping_cost: number | null
  is_primary: boolean
  created_at: string
}

export interface Part {
  id: string
  build_id: string
  name: string
  description: string | null
  category: string | null
  status: "needed" | "ordered" | "sourced" | "installed"
  price_estimate: number | null
  vendor_url: string | null
  image_url: string | null
  is_safety_critical: boolean
  notes: string | null
  goal: string | null
  vendors: PartVendor[]
  ordered_from_vendor_id: string | null
  ordered_at: string | null
  created_at: string
  updated_at: string
}

export interface VisionExtracted {
  make: string | null
  model: string | null
  year: string | null
  confidence: number | null
  part_name: string | null
  specifications: Record<string, string> | null
  mods_detected: string[]
  notes: string | null
}

export interface VisionData {
  image_type: 'car' | 'rims' | 'engine_bay' | 'suspension' | 'inspiration' | 'part' | 'unknown'
  summary: string
  extracted: VisionExtracted
}

export interface BuildDetail extends Omit<Build, "vision_data"> {
  parts: Part[]
  vision_data: VisionData | null
}

export interface GenerateResponse {
  parts_created: number
  build: BuildDetail
}

export interface GeneratePartsResponse {
  build_id: string
  parts: Part[]
  total_parts: number
  estimated_total: number
  safety_critical_count: number
  message: string
}

export interface CreateBuildPayload {
  title: string
  car?: string
  modification_goal?: string
  goals: string[]
}

export interface BuildsResponse {
  builds: Build[]
  totalParts: number
  sourcedParts: number
  estimatedSpend: number
}

const CLIENT_API_URL = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"

export async function createBuild(
  payload: CreateBuildPayload,
  accessToken: string
): Promise<Build> {
  const res = await fetch(`${CLIENT_API_URL}/v1/builds/`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${accessToken}`,
    },
    body: JSON.stringify(payload),
  })

  if (!res.ok) {
    const error = await res.json().catch(() => ({ detail: "Unknown error" }))
    throw new Error(error.detail ?? `API error ${res.status}`)
  }

  return res.json()
}

export async function updateBuild(
  id: string,
  payload: Partial<Pick<Build, "title" | "car" | "modification_goal" | "goals" | "status">>,
  accessToken: string
): Promise<BuildDetail> {
  const res = await fetch(`${CLIENT_API_URL}/v1/builds/${id}`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${accessToken}`,
    },
    body: JSON.stringify(payload),
  })

  if (!res.ok) {
    const error = await res.json().catch(() => ({ detail: "Unknown error" }))
    throw new Error(error.detail ?? `API error ${res.status}`)
  }

  return res.json()
}

export async function generateParts(
  buildId: string,
  accessToken: string
): Promise<GenerateResponse> {
  const res = await fetch(`${CLIENT_API_URL}/v1/builds/${buildId}/generate`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${accessToken}`,
    },
  })

  if (!res.ok) {
    const error = await res.json().catch(() => ({ detail: "Unknown error" }))
    throw new Error(error.detail ?? `API error ${res.status}`)
  }

  return res.json()
}

export async function generatePartsNew(
  buildId: string,
  accessToken: string,
  forceRegenerate: boolean = false
): Promise<GeneratePartsResponse> {
  const res = await fetch(`${CLIENT_API_URL}/v1/builds/${buildId}/parts/generate`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${accessToken}`,
    },
    body: JSON.stringify({ force_regenerate: forceRegenerate }),
  })

  if (!res.ok) {
    const error = await res.json().catch(() => ({ detail: "Unknown error" }))
    throw new Error(error.detail ?? `API error ${res.status}`)
  }

  return res.json()
}

export async function uploadBuildImage(
  buildId: string,
  image: File,
  accessToken: string
): Promise<{ image_url: string }> {
  const form = new FormData()
  form.append("image", image)

  const res = await fetch(`${CLIENT_API_URL}/v1/builds/${buildId}/image`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${accessToken}`,
    },
    body: form,
  })

  if (!res.ok) {
    const error = await res.json().catch(() => ({ detail: "Unknown error" }))
    throw new Error(error.detail ?? `API error ${res.status}`)
  }

  return res.json()
}

export async function orderPart(
  buildId: string,
  partId: string,
  vendorId: string,
  accessToken: string
): Promise<Part> {
  const res = await fetch(`${CLIENT_API_URL}/v1/builds/${buildId}/parts/${partId}/order`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${accessToken}`,
    },
    body: JSON.stringify({ vendor_id: vendorId }),
  })

  if (!res.ok) {
    const error = await res.json().catch(() => ({ detail: "Unknown error" }))
    throw new Error(error.detail ?? `API error ${res.status}`)
  }

  return res.json()
}
